### PR TITLE
Qr cnc svg

### PR DESCRIPTION
**Pull request by spatcho**
### What is this PR for?

Print qr to cnc, write to file : adapt code to latest qr_code array structure

Print qr to cnc, write directly to cnc  : uncomment, test, and adapt code to fit an openbuilds grbl 1.1 cnc, see comments on cnc.py

Save qr as svg image

### Changes made to:
- [X] Code
- [ ] Tests
- [ ] Docs
- [X] CHANGELOG

### What is the purpose of this pull request?
- [X] Bug fix
- [X] New feature
- [ ] Docs update
- [ ] Other

### CODE REVIEW

**Comment by qlrd** [comment]
@spatcho, reviewed a little your changes and there're some broken tests:

```bash
FAILED tests/printers/test_cnc.py::test_print_qr_code_with_row_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
FAILED tests/printers/test_cnc.py::test_print_qr_code_with_spiral_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
FAILED tests/printers/test_cnc.py::test_print_qr_code_inverted - TypeError: unsupported operand type(s) for &: 'str' and 'int'
```

I would be nice if you can be more specific in the PR about the changes (i will be very happy to assist you on that).

File: src/krux/printers/cnc.py [hunk 1]
```diff
@@ -289,6 +289,7 @@ def clear(self):
 # On krux device, the grbl/cnc printer driver need to be selected, here is the settings tested on : {"settings": {"persist": {"location": "sd"}, "printer": {"driver": "cnc/file", "cnc": {"unit": "mm", "part_size": 70.675, "flute_diameter": 3.175, "depth_per_pass": 1.0, "cut_depth": 2.0, "border_padding": 2.0, "plunge_rate":300, "feed_rate":650,"cut_method": "spiral"}}}}
 # It seems the wondermv device can be powered by the blackbox controller only but sometimes it doesn't start. I it's usb powered it always start.
 # Testing scenario : power the cnc, use the interface to home everything and start the router, unplug the interface and plug the krux device instead, start print.
+# Note : This has been tested on the machine without router started and without bit, only to see that it move correctly and that it complete a full cycle.
```
Comment (qlrd) [inline src/krux/printers/cnc.py:292]: I think this change could be squashed with previous commit

File: mkdocs.yml [hunk 2]
```diff
@@ -51,7 +51,7 @@ edit_uri: edit/main/docs
 docs_dir: docs
 site_dir: public
 extra:
-  latest_krux: krux-v25.03.0
+  latest_krux: krux-v25.04.0
```
Comment (qlrd) [inline mkdocs.yml:54]: This change is unrelated with the specifics of this commit. 

Maybe this change can be done by @odudex when the release is done (i believe you can discard it)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,9 @@
+# Changelog 25.04.0 - April 2025
```
Comment (qlrd) [inline CHANGELOG.md:1]: This change is unrelated with the specifics of this commit. 

Maybe this change can be done by @odudex when the release is done.

Anyways, it's worth to keep the information (but i believe when the `develop` branch will go to `main`, they will appear automatically).

File: pyproject.toml [hunk 4]
```diff
@@ -22,7 +22,7 @@
 
 [tool.poetry]
 name = "krux"
-version = "25.03.0"
+version = "25.04.0"
```
Comment (qlrd) [inline pyproject.toml:25]: This change is unrelated with the specifics of this commit. 

Maybe this change can be done by @odudex when the release is done (i believe you can discard it)

File: src/krux/metadata.py [hunk 5]
```diff
@@ -19,5 +19,5 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
-VERSION = "25.03.0"
+VERSION = "25.04.0"
```
Comment (qlrd) [inline src/krux/metadata.py:22]: This change is unrelated with the specifics of this commit. 

Maybe this change can be done by @odudex when the release is done (i believe you can discard it)

File: src/krux/pages/qr_view.py [hunk 6]
```diff
@@ -384,6 +384,45 @@ def save_bmp_image(self, file_name, resolution):
         bmp_img.save("/sd/" + file_name)
         self.flash_text(t("Saved to SD card") + ":\n%s" % file_name)
 
+    def save_svg_image(self, file_name):
+        """Save QR code image as .svg file"""
+        from ..sd_card import SVG_IMAGE_EXTENSION
+        from .file_operations import SaveFile
+
+        import image
+        import lcd
```
Comment (qlrd) [inline src/krux/pages/qr_view.py:393]: IMO, these imports could be togheter with other imports at the top of the file.

File: src/krux/printers/cnc.py [hunk 7]
```diff
@@ -68,12 +68,19 @@ def qr_data_width(self):
 
     def print_qr_code(self, qr_code):
         """Prints a QR code, scaling it up as large as possible"""
+        from ..qr import get_size
```
Comment (qlrd) [inline src/krux/printers/cnc.py:71]: IMO, these imports could be togheter with other imports at the top of the file.

File: src/krux/printers/cnc.py [hunk 8]
```diff
@@ -83,26 +90,35 @@ def print_qr_code(self, qr_code):
...
                         # Flip the y coord
+                        y = row
+                        x = col
+                        x_index = x
+                        plunge_depth = min((p + 1) * self.pass_depth, self.cut_depth)
                         self.cut_cell(x_index, size - 1 - y, cell_size, plunge_depth)
 
     def cut_cell(self, x, y, cell_size, plunge_depth):
         """Hollows out the specified cell using a cutting method defined in settings"""
+        #print("cut_cell:",x,",",y,",",cell_size,",",plunge_depth)
```
Comment (qlrd) [inline src/krux/printers/cnc.py:121]: It's necessary to keep the comment (maybe a hint for you or others)?

If not, it's worth to remove it.

File: src/krux/printers/cnc.py [hunk 9]
```diff
@@ -268,54 +284,108 @@ def clear(self):
...
-#             raise ValueError("gcode send failed: timed out")
-
-#         status = res.decode().split("\n")[0]
-#         if status != "ok":
-#             err_msg = status if status.startswith("error") else "unknown error"
-#             raise ValueError("gcode send failed: %s" % err_msg)
-
-#     def clear(self):
-#         """Clears the printer's memory, resetting it"""
+# Tested on openbuilds 1515 with openbuilds blackbox x4 grbl controller, openbuilds interface serial remote controller, and wondermv.
```
Comment (qlrd) [inline src/krux/printers/cnc.py:287]: IMO, it's worth to change this comment into an docstrings, so editors like vscode and nvim could show these docstrings.

Example:

```python
# An adapter was made to use the same cable that connect to interface, 
# with rx and tx reversed. The machine need be homed first before the krux 
# software send the commands.
...
```

Could be:

```python
"""
An adapter was made to use the same cable that connect to interface, 
with rx and tx reversed. The machine need be homed first before the krux 
software send the commands.
...
"""
```

File: src/krux/printers/cnc.py [hunk 10]
```diff
@@ -83,26 +90,35 @@ def print_qr_code(self, qr_code):
...
+            #print("pass #",p)
+            for row in range(size):
+                #print("row #",row)
+                byte = 0
+                line_bytes = bytearray()
+                for col in range(size):
+                    #print("col #",col)
+                    bit_index = row * size + col
+                    #print("bit_index #",bit_index)
+                    bit = qr_code[bit_index >> 3] & (1 << (bit_index % 8))
```
Comment (qlrd) [inline src/krux/printers/cnc.py:106]: Here occurs the same error for 3 tests (`test_print_qr_code_with_row_cutmethod`, 
`test_print_qr_code_with_spiral_cutmethod` and `test_print_qr_code_inverted`):

```bash
self = <krux.printers.cnc.FilePrinter object at 0x106a39ac0>
qr_code = '00000000000000000000000000000000000\n01111111001110001000100110011111110\n01000001010011001001011110010000010\n010111...1100110\n01000001010010000111101011001110000\n01111111010000110100011011010110100\n00000000000000000000000000000000000'

    def print_qr_code(self, qr_code):
        """Prints a QR code, scaling it up as large as possible"""
        from ..qr import get_size

        size = 0

        size = get_size(qr_code)
        print("size:",size)

        print("self.part_size",self.part_size," / self.border_padding:",self.border_padding," / size:",size)

        cell_size = (self.part_size - (self.border_padding * 2)) / size

        print("cell_size:",cell_size)

        # Modal settings
        self.on_gcode("G17")  # x/y plane
        self.on_gcode("G20" if self.unit == "in" else "G21")  # units
        self.on_gcode("G40")  # cancel diameter compensation
        self.on_gcode("G49")  # cancel length offset
        self.on_gcode("G54")  # coord system 1
        self.on_gcode("G90")  # non-incremental motion
        self.on_gcode("G94")  # feed/minute mode

        print("self.cut_depth:",self.cut_depth," / self.pass_depth:",self.pass_depth)

        num_passes = math.ceil(self.cut_depth / self.pass_depth)
        for p in range(num_passes):
            #print("pass #",p)
            for row in range(size):
                #print("row #",row)
                byte = 0
                line_bytes = bytearray()
                for col in range(size):
                    #print("col #",col)
                    bit_index = row * size + col
                    #print("bit_index #",bit_index)
>                   bit = qr_code[bit_index >> 3] & (1 << (bit_index % 8))
E                   TypeError: unsupported operand type(s) for &: 'str' and 'int'

src/krux/printers/cnc.py:106: TypeError
```

**Review by qlrd — CHANGES_REQUESTED** [review]

File: src/krux/printers/cnc.py (hunk 10 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:106]: The test need to be adapted because the qr_code is no more a string with 0, 1 and \n.

File: src/krux/printers/cnc.py (hunk 8 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:121]: What would be best would be that it print in debug mode and not otherwise, because if I remove and then I have to debug again, I have to put all the prints back. Does this project support such features ?

File: src/krux/printers/cnc.py (hunk 1 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:292]: This has been real life tested now. You want me to remove the line and push again ?

File: src/krux/printers/cnc.py (hunk 7 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:71]: I don't agree personally but no problem to make the change, shall I ?

File: src/krux/pages/qr_view.py (hunk 6 above)
Comment (spatcho) [reply src/krux/pages/qr_view.py:393]: I don't agree personally but no problem to make the change, shall I ?

File: pyproject.toml (hunk 4 above)
Comment (spatcho) [reply pyproject.toml:25]: I wanted a different version to know which firmware is running on the device. No problem to ignore it. Shall I do something here ?

File: src/krux/metadata.py (hunk 5 above)
Comment (spatcho) [reply src/krux/metadata.py:22]: I wanted a different version to know which firmware is running on the device. No problem to ignore it. Shall I do something here ?

File: CHANGELOG.md (hunk 3 above)
Comment (spatcho) [reply CHANGELOG.md:1]: That was the logical place for me to put the comments on what was done. No problem to ignore it. Shall I do something here ?

File: mkdocs.yml (hunk 2 above)
Comment (spatcho) [reply mkdocs.yml:54]: I think this is generated and I should not have push it. No problem to ignore it. Shall I do something here ?

File: src/krux/printers/cnc.py (hunk 9 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:287]: Will see to edit it in vscode and push an update.

**Comment by spatcho** [comment]
> @spatcho, reviewed a little your changes and there're some broken tests:
> 
> ```shell
> FAILED tests/printers/test_cnc.py::test_print_qr_code_with_row_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> FAILED tests/printers/test_cnc.py::test_print_qr_code_with_spiral_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> FAILED tests/printers/test_cnc.py::test_print_qr_code_inverted - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> ```
> 
> I would be nice if you can be more specific in the PR about the changes.

How much more specific ?

**Comment by tadeubas** [comment]
Hey @spatcho you are aiming MASTER branch, it is wrong, we first aim for DEVELOP then later in time we merge DEVELOP into MASTER creating a new version, this is why you didn't see other changes on `CHANGELOG` file for example.

Please just **edit** you PR and change the branch like this one in the img below:
image

File: src/krux/pages/qr_view.py (hunk 6 above)
Comment (tadeubas) [reply src/krux/pages/qr_view.py:393]: Hey @qlrd sometimes we prefer to import things inside the function in order to avoid unnecessary import when not using the lib everywhere, so I don't see any problem here as it appears that he followed other already implemented methods like `save_pbm_image` and `save_bmp_image`

File: .gitignore [hunk 11]
```diff
@@ -84,3 +84,6 @@ krux-*/ktool*
 
 # IDE files
 .vscode
+ktool.py
+src/krux/metadata.py--
+mkdocs.yml--
```
Comment (tadeubas) [inline .gitignore:89]: Why add these two values with `--` at the end?

Comment (qlrd) [reply .gitignore:89]: Sometime ago some files with same -- was generated when i built the firmware

File: CHANGELOG.md (hunk 3 above)
Comment (tadeubas) [reply CHANGELOG.md:1]: When you aim for develop and pull / merge to your branch you will see we are currently using `# Changelog 25.XX.X - XXX 2025`

File: src/krux/printers/cnc.py (hunk 10 above)
Comment (qlrd) [reply src/krux/printers/cnc.py:106]: you can run all with `poetry run poe test` (good to see coverage) or if you just want to test the features `poetry run pytest tests/<the-path-to-be-tested>.py`.

File: src/krux/metadata.py (hunk 5 above)
Comment (tadeubas) [reply src/krux/metadata.py:22]: In develop we use beta version, like `VERSION = "25.05.beta0"`, but it is always better to not commit these changes as it is something related only to you and your tests, @odudex sometimes releases new beta versions to people to test what is currently in our develop branch, so at this time he commits and increments the beta to `25.05.beta1` and so on...

File: src/krux/printers/cnc.py (hunk 7 above)
Comment (tadeubas) [reply src/krux/printers/cnc.py:71]: Hey @qlrd , as I mentioned above it is ok to do this inside a function because it can decrease the memory usage by only importing this file when the function is used. If it was something used by this file everywhere than it would be better to import at the top (these are little optimizations because we deal with limited memory, it does not apply to normal software development)

File: src/krux/printers/cnc.py [hunk 12]
```diff
@@ -68,12 +68,19 @@ def qr_data_width(self):
...
     def print_qr_code(self, qr_code):
         """Prints a QR code, scaling it up as large as possible"""
+        from ..qr import get_size
+
         size = 0
-        while qr_code[size] != "\n":
-            size += 1
+
+        size = get_size(qr_code)
+        print("size:",size)
```
Comment (tadeubas) [inline src/krux/printers/cnc.py:76]: Hey @spatcho these `prints` were used for debug right? If so, you should remove them as it is not good to be printing unnecessary things in terminal.

File: src/krux/printers/cnc.py (hunk 9 above)
Comment (qlrd) [reply src/krux/printers/cnc.py:287]: You can do:

- `git add <file>` (change the files);
- `git commit` (update the history);
- `git rebase -i HEAD~2` (you get the latest commit plus the current). An editor will be show:

```bash
pick abc123 My latest commit
pick cde456 My current commit
```

Then change to it

```bash
pick abc123 My latest commit
squash cde456 My current commit
```

The changes you do will be merged like one. Then to push:

- `git push --force-with-lease <your-fork> <your-branch>`

If you prefer, we can do together when you have time.

File: src/krux/printers/cnc.py (hunk 1 above)
Comment (qlrd) [reply src/krux/printers/cnc.py:292]: You can do:

- `git add <file>` (change the files);
- `git commit` (update the history);
- `git rebase -i HEAD~2` (you get the latest commit plus the current). An editor will be show:

```bash
pick abc123 My latest commit
pick cde456 My current commit
```

Then change to it

```bash
pick abc123 My latest commit
squash cde456 My current commit
```

The changes you do will be merged like one. Then to push:

- `git push --force-with-lease <your-fork> <your-branch>`

If you prefer, we can do together when you have time.

File: mkdocs.yml (hunk 2 above)
Comment (qlrd) [reply mkdocs.yml:54]: Understood, it happens :)

File: CHANGELOG.md (hunk 3 above)
Comment (qlrd) [reply CHANGELOG.md:1]: There's a way to remove those files from commit. We can do togheter when you have time.

**Comment by qlrd** [comment]
> > @spatcho, reviewed a little your changes and there're some broken tests:
> > ```shell
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_row_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_spiral_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_inverted - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I would be nice if you can be more specific in the PR about the changes.
> 
> How much more specific ?



> > @spatcho, reviewed a little your changes and there're some broken tests:
> > ```shell
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_row_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_spiral_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > FAILED tests/printers/test_cnc.py::test_print_qr_code_inverted - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > ```
> > I would be nice if you can be more specific in the PR about the changes.
> 
> How much more specific ?

I advocate for verbosity. No need to explain like im 5 years old, but i'm trying to understand the firmware too (i came from installer and studying firmware to help with tests). 

So i think if (new potentially) developers could know your rationale would be a great plus.

File: src/krux/printers/cnc.py (hunk 10 above)
Comment (tadeubas) [reply src/krux/printers/cnc.py:106]: We use poetry to run the project and tests, you can see some instructions on our `README`, some tasks are not listed there like `poetry run poe pre-commit` that will run all develop related tasks (needed to be done before a commit or before creating a PR). Take a look at the `pyproject.toml` file to see all tasks.

Currently we are recommending VSCode and to use their test tab as it is easier to run a specific test only. See the img below:

image

File: src/krux/printers/cnc.py (hunk 8 above)
Comment (tadeubas) [reply src/krux/printers/cnc.py:121]: It supported something like this, but we removed because it is unnecessary code just needed for development. When development finishes and the changes are ready to a PR, you should remove these prints or comments. If the PR is still in progress and development, just change the PR on Github to draft and when done with everything click on the button that appears "Ready for review"

File: pyproject.toml (hunk 4 above)
Comment (qlrd) [reply pyproject.toml:25]: I think to check in device is ok. I think just remove the change on commit will do

File: src/krux/metadata.py (hunk 5 above)
Comment (qlrd) [reply src/krux/metadata.py:22]: I think just patching the commit and then remove it, is ok

File: src/krux/printers/cnc.py (hunk 12 above)
Comment (spatcho) [reply src/krux/printers/cnc.py:76]: Yes that was noticed previously.

**Comment by spatcho** [comment]
> Hey @spatcho you are aiming MASTER / MAIN branch, it is wrong, we first aim for DEVELOP then later in time we merge DEVELOP into MASTER / MAIN creating a new version, this is why you didn't see other changes on `CHANGELOG` file for example.
> 
> Please just **edit** you PR and change the branch like this one in the img below: image

It's done, you confirm ?

**Comment by qlrd** [comment]
> > > > @spatcho, reviewed a little your changes and there're some broken tests:
> > > > ```shell
> > > > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_row_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > > > FAILED tests/printers/test_cnc.py::test_print_qr_code_with_spiral_cutmethod - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > > > FAILED tests/printers/test_cnc.py::test_print_qr_code_inverted - TypeError: unsupported operand type(s) for &: 'str' and 'int'
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I would be nice if you can be more specific in the PR about the changes.
> > > 
> > > 
> > > How much more specific ?
> > 
> > 
> > I advocate for verbosity. No need to explain like im 5 years old, but i'm trying to understand the firmware too (i came from installer and studying firmware to help with tests).
> > So i think if (new potentially) developers could know your rationale would be a great plus.
> 
> Please have a look now.

Amazing, now i understood what you want/ and it's great!

File: mkdocs.yml (hunk 2 above)
Comment (tadeubas) [reply mkdocs.yml:54]: You can just edit this file again returning it to the older value (unchanged)

File: CHANGELOG.md (hunk 3 above)
Comment (tadeubas) [reply CHANGELOG.md:1]: You have to edit this file and add your changes with our under the `25.XX.X`, then remove this line `25.04.0` you've creater

File: src/krux/metadata.py (hunk 5 above)
Comment (tadeubas) [reply src/krux/metadata.py:22]: You can just edit this file again returning it to the older value (unchanged)

**Comment by odudex** [comment]
Time to refine the PR
run `poetry run poe black` to format your code changes.
run `poetry run poe lint` to run pylint to check some code compliance parameters, then you have to address the issues pointed by it.
run `poetry run poe test` to run our automated tests. Each broken test must be evaluated, sometimes the code is flawed, other times the test itself must be adapted.
Finally, there will be a test coverage test, where we ensure at least 90% of new code is covered by tests.

**Review by qlrd — APPROVED** [review]
tACK 0bc7ac23d66331862ddfc4ed65028e08f6aae890

The coverage is 94%, but the uncovered code happens in exceptions. 

I can work in these edge cases if you guys agree (since I set out to do this work 😁)

**Comment by qlrd** [comment]
nit: the coverage is 94%, but happens in exceptions. 

I can work in these edge cases if you guys agree.

**Comment by odudex** [comment]
Ok, now that I merged other pull requests, this branch has some conflicts to be solved. The PR also have some changes in unrelated files.
Since this PR is also for learning (it's @spatcho 's first PR, and we're also learning how to manage multiple development fronts), I would suggest to restart the PR from scratch, or better, create 2 PRs, one at a time.
The code, of course, should be reused, @spatcho can make a backup elsewhere, then do the following.
Fist, let's do a simpler PR, I believe it would be the .svg export
Check out to develop branch, run `git fetch origin`, `git checkout develop`, `git submodule update --recursive` and ensure your git tree is clean.
Create a new branch, like `svg_export`, and
 - Insert the code required to make svg export to work.
 - Edit changelog and add respective information.
   (We can help you stage and commit only changes in files that are related to the PR)
 - Run `poetry run poe black`, `poetry run poe lint`, `poetry run poe test`. We'll help with any issue that you might have.
 - Only after everything is green, we do a final review and merge the "SVG Export" PR.

After this we do the same with a "CNC Fix and Update"

What do you think?

**Comment by qlrd** [comment]
Agreed, there's a lot of ways to manipulate the git history so he do not lost his work. 

I think with some `git add --patch`, `rebase`, `fix` and `drop`s is possible to separate this in two PRs without type any code

**Comment by tadeubas** [comment]
> Ok, now that I merged other pull requests, this branch has some conflicts to be solved. The PR also have some changes in unrelated files. Since this PR is also for learning (it's @spatcho 's first PR, and we're also learning how to manage multiple development fronts), I would suggest to restart the PR from scratch, or better, create 2 PRs, one at a time. The code, of course, should be reused, @spatcho can make a backup elsewhere, then do the following. Fist, let's do a simpler PR, I believe it would be the .svg export Check out to develop branch, run `git fetch origin`, `git checkout develop`, `git submodule update --recursive` and ensure your git tree is clean. Create a new branch, like `svg_export`, and
> 
> * Insert the code required to make svg export to work.
> * Edit changelog and add respective information.
>   (We can help you stage and commit only changes in files that are related to the PR)
> * Run `poetry run poe black`, `poetry run poe lint`, `poetry run poe test`. We'll help with any issue that you might have.
> * Only after everything is green, we do a final review and merge the "SVG Export" PR.
> 
> After this we do the same with a "CNC Fix and Update"
> 
> What do you think?

I like to use the `poetry run poe pre-commit` as it executes the `black, lint, check for new translations and tests` all in one simple command. 

IMO it is simpler to just do the merge, he already removed the unnecessary changes to unrelated things. If @spatcho wants I can resolve the conflicts and create a PR to his branch in its fork and when accepted this PR here will be automatically updated.

**Comment by spatcho** [comment]
> > Ok, now that I merged other pull requests, this branch has some conflicts to be solved. The PR also have some changes in unrelated files. Since this PR is also for learning (it's @spatcho 's first PR, and we're also learning how to manage multiple development fronts), I would suggest to restart the PR from scratch, or better, create 2 PRs, one at a time. The code, of course, should be reused, @spatcho can make a backup elsewhere, then do the following. Fist, let's do a simpler PR, I believe it would be the .svg export Check out to develop branch, run `git fetch origin`, `git checkout develop`, `git submodule update --recursive` and ensure your git tree is clean. Create a new branch, like `svg_export`, and
> > 
> > * Insert the code required to make svg export to work.
> > * Edit changelog and add respective information.
> >   (We can help you stage and commit only changes in files that are related to the PR)
> > * Run `poetry run poe black`, `poetry run poe lint`, `poetry run poe test`. We'll help with any issue that you might have.
> > * Only after everything is green, we do a final review and merge the "SVG Export" PR.
> > 
> > After this we do the same with a "CNC Fix and Update"
> > What do you think?
> 
> I like to use the `poetry run poe pre-commit` as it executes the `black, lint, check for new translations and tests` all in one simple command.
> 
> IMO it is simpler to just do the merge, he already removed the unnecessary changes to unrelated things. If @spatcho wants I can resolve the conflicts and create a PR to his branch in its fork and when accepted this PR here will be automatically updated.

@tadeubas if you could do the resolve as proposed I would much appreciate it

**Comment by odudex** [comment]
That works too!
I still see changes in .gitignore and PR templates that needs to be undone.

File: .gitignore (hunk 11 above)
Comment (spatcho) [reply .gitignore:89]: I build with ./krux build maixpy_wonder_mv , it's docker afaik . I want to keep using docker. I am not using git on the command line, I use a visual tool (Fork app), and I don't like to see those files that actually need to be ignored. I also added *.code-workspace and */.DS_Store because I am a Mac and it annoyingly put those DS_Store files .

**Comment by spatcho** [comment]
> That works too! I still see changes in .gitignore and PR templates that needs to be undone.

About .gitignore : I replied to your comment

About PR templates : what is it ?

File: .gitignore (hunk 11 above)
Comment (qlrd) [reply .gitignore:89]: Maybe it's worth to put these files in `.gitignore`, but in another PR (to keep the history review managable). 

I believe these rules could be something like:

```diff
@@ -84,3 +84,6 @@ krux-*/ktool*

# IDE files
.vscode
+ ktool*
+ **/*.*--
+ **/.DS_Store
```

**Comment by qlrd** [comment]
> > That works too! I still see changes in .gitignore and PR templates that needs to be undone.
> 
> About .gitignore : I replied to qlrd comment here [#557 (comment)](https://github.com/selfcustody/krux/pull/557#discussion_r2033883678) , is it about something else ?
> 
> About PR templates : what is it ?

PR/issue templates are files that are used when someone opens a PR/issues. 

You remember when you created this PR and the related issue and had some pre-filled text? it's that.

**Comment by spatcho** [comment]
I pushed a reverted .gitignore. I will try to keep mine unpushed. I want to close this PR.

**Comment by spatcho** [comment]
> > > That works too! I still see changes in .gitignore and PR templates that needs to be undone.
> > 
> > 
> > About .gitignore : I replied to qlrd comment here [#557 (comment)](https://github.com/selfcustody/krux/pull/557#discussion_r2033883678) , is it about something else ?
> > About PR templates : what is it ?
> 
> PR/issue templates are files that are used when someone opens a PR/issues.
> 
> You remember when you created this PR and the related issue and had some pre-filled text? it's that.

And what shall I undone ?

File: .gitignore (hunk 11 above)
Comment (spatcho) [reply .gitignore]: It has been reverted so case closed

**Comment by qlrd** [comment]
> That works too! I still see changes in .gitignore and PR templates that needs to be undone.

I believe it's worth to a rebase too, see that are two unrelated commits from @tadeubas and one from @odudex. You have telegram or discord? maybe we can do togheter (with @tadeubas too).

**Comment by qlrd** [comment]
> > > > That works too! I still see changes in .gitignore and PR templates that needs to be undone.
> > > 
> > > 
> > > About .gitignore : I replied to qlrd comment here [#557 (comment)](https://github.com/selfcustody/krux/pull/557#discussion_r2033883678) , is it about something else ?
> > > About PR templates : what is it ?
> > 
> > 
> > PR/issue templates are files that are used when someone opens a PR/issues.
> > You remember when you created this PR and the related issue and had some pre-filled text? it's that.
> 
> And what shall I undone ?
> 
> You mean that ? Screenshot 2025-04-09 at 14 01 08

A rebase resolve that, if you wanna, find-me on Krux group and lets talk

**Comment by spatcho** [comment]
I can send you or you can extract patch files from my work on include it in the project.

**Comment by odudex** [comment]
Loop talks and focus deviations are exactly what I would like to avoid. For that, I think the best strategy is to keep it simple, break things that are starting to fold on itself in small peaces, finish, and move them behind. Not introducing any new topic, except minimum requirements. The goal is simple, to have the code fully compliant. Let me know if you need any help.

**Comment by tadeubas** [comment]
Hi guys, I believe the changes in `.github/PULL_REQUEST_TEMPLATE.md` are just because it started in `main` and is now in `develop`, so we changed that template and forgot to merge `main` into `develop`, but Git is smart and detects and merges the changes later, and if there is a conflict we can always merge manually too.

We have had a lot of PRs changing `.gitignore`, it is not a file that needs to be optimized, it is not integrated into the firmware, it is just a file to help developers. I was just wondering what the additions were and if they were necessary, of course we should add `*.code-workspace` (and any other IDE specific files) and `*/.DS_Store` (for Mac users).

@spatcho has already said he does not use Git on the command line, Git is not easy and can be a hassle and a burden. He just wants to close this PR (meaning he wants this merged). It will go to `develop`, so I guess we can iterate on the code later, since we are more familiar with the code being fully compliant and it is clear to me that these requests are taking the joy out of coding for the new contributor.

I will merge it with rebase (with the help of @qlrd of course, because I am not used to it). Afterwards, I promise to add something basic to the documentation mentioning this feature. After a long time living in the firmware, this CNC feature has been really tested and improved. Thanks @spatcho for your work and testing! I hope you continue contributing to Krux in the near future! I will create the PR to your fork and when you accept this PR here will be automatically updated.

**Comment by odudex** [comment]
> so we changed that template and forgot to merge `main` into `develop`

Good catch. Merged `main` into `develop`

**Comment by qlrd** [comment]
@spatcho, you work is awesome. I think you will be a great addition!

I know git could be hard at first, and was for everyone on the team. I hope you can understand what we want to achieve is simplicity and readability on reviews.

> Hi guys, I believe the changes in `.github/PULL_REQUEST_TEMPLATE.md` are just because it started in `main` and is now in `develop`, so we changed that template and forgot to merge `main` into `develop`, but Git is smart and detects and merges the changes later, and if there is a conflict we can always merge manually too.
> 
> We have had a lot of PRs changing `.gitignore`, it is not a file that needs to be optimized, it is not integrated into the firmware, it is just a file to help developers. I was just wondering what the additions were and if they were necessary, of course we should add `*.code-workspace` (and any other IDE specific files) and `*/.DS_Store` (for Mac users).
> 
> @spatcho has already said he does not use Git on the command line, Git is not easy and can be a hassle and a burden. He just wants to close this PR (meaning he wants this merged). It will go to `develop`, so I guess we can iterate on the code later, since we are more familiar with the code being fully compliant and it is clear to me that these requests are taking the joy out of coding for the new contributor.
> 
> I will merge it with rebase (with the help of @qlrd of course, because I am not used to it). Afterwards, I promise to add something basic to the documentation mentioning this feature. After a long time living in the firmware, this CNC feature has been really tested and improved. Thanks @spatcho for your work and testing! I hope you continue contributing to Krux in the near future! I will create the PR to your fork and when you accept this PR here will be automatically updated.

Yeah, this is what happened. Rebase is a bad idea althought (i tried here and had 98 conflicts).

I think a better approach is branch a `qr_cnc` and a `qr_svg` from `develop` by `reset`ing it  to current state of `develop`, start to `cherry-pick` each one of the commits, resolve conflicts one by one (3 commits, i will need the help of @tadeubas to select which changes are the correct) and `add --patch`/ `stash` to separate the PRs.

So we will not lost @spatcho awesome work neither his authoring.

**Comment by qlrd** [comment]
I think i managed to do what i said.

Please review it: https://github.com/selfcustody/krux/compare/develop...qlrd:krux:qr_cnc?expand=1

@spatcho, if i didnt missed anything in CNC stuff, i will send here a `diff`, and you can apply it to a new PR.

**Comment by qlrd** [comment]
@spatcho, i think i managed to separate the CNC code. Could you review to see if i didnt forget anything?

https://github.com/selfcustody/krux/compare/develop...qlrd:krux:qr_cnc?expand=1

**Comment by qlrd** [comment]
@spatcho, can you review and test https://github.com/selfcustody/krux/pull/562 ?

**Comment by qlrd** [comment]
@tadeubas, maybe the new laser adaption be in the PR #562 or be in another separated PR?

**Comment by odudex** [comment]
Is there something from here yet to be implemented in another PR? Or can we close it?

**Comment by tadeubas** [comment]
Laser CNC is still missing a PR

**Comment by qlrd** [comment]
> Laser CNC is still missing a PR

Just started it

**Comment by qlrd** [comment]
Since #562, #585, #590 and #593 was merged, i think this could be closed.

**Comment by odudex** [comment]
> Since #562, #585, #590 and #593 was merged, i think this could be closed.

Were all features from above PRs tested with real devices(CNCs, laser, etc)?
